### PR TITLE
Passing Config and Parser information through TextWrapper

### PR DESCRIPTION
**Pull request by alexbowers**

### CODE REVIEW

**Comment by alexbowers** [comment]
#433 Linking for reference

**Comment by alexbowers** [comment]
Example Code Usage:

<?php
    $app = new \Phinx\Console\PhinxApplication();
    $wrap = new \Phinx\Wrapper\TextWrapper($app, array(
                'config_path' => '../phinx.yml',
                'parser' => 'yaml'
        ));
    $test = $wrap->getMigrate();

    print_r($test);
It works without the second parameter. 
Either, all or none of the param values are required and can be passed. parser is an assumed value based on extension of the config_path, default config path is ./phinx.yml, Also, you can pass through the environment.

File: src/Phinx/Config/Config.php [hunk 1]
```diff
@@ -49,6 +49,11 @@ class Config implements ConfigInterface
     protected $configFilePath;
 
     /**
+     * @var $phinx_version
+     */
+    public static $phinx_version = "0.4.1";
```
Comment (shadowhand) [inline src/Phinx/Config/Config.php]: Why is this added here? We already have a `PHINX_VERSION` constant.

File: src/Phinx/Console/PhinxApplication.php [hunk 2]
```diff
@@ -47,8 +48,12 @@ class PhinxApplication extends Application
      *
      * @param string $version The Application Version
      */
-    public function __construct($version)
+    public function __construct($version = null)
```
Comment (shadowhand) [inline src/Phinx/Console/PhinxApplication.php]: This is totally unrelated to the PR title/description. Please revert it.

File: src/Phinx/Wrapper/TextWrapper.php [hunk 3]
```diff
@@ -42,17 +42,20 @@ class TextWrapper
...
-     * @return void
+     * @param PhinxApplication $app
+     * @param array $options
      */
-    public function __construct(PhinxApplication $app, $default_env = 'development')
+    public function __construct(PhinxApplication $app, Array $options = array())
     {
         $this->app = $app;
-        $this->env = $default_env;
+        $this->env = (isset($options['environment']) ? $options['environment'] : 'development');
```
Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: Why limit the access to options by assigning them in the constructor? Wouldn't it be better to store **all** the options and have a `getOption` method, like in `AdapterInterface`?

File: src/Phinx/Wrapper/TextWrapper.php [hunk 4]
```diff
@@ -107,6 +110,45 @@ public function getRollback($env = null, $target = null)
         return $this->executeRun($command);
     }
 
+    /**
+     * Set the environment
+     *
+     * @param string $environment
+     */
+    public function setEnvironment($environment = null) {
```
Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: What's the reason for these? Nothing seems to use them, and they can already be modified by passing the argument to the specific method.

File: src/Phinx/Wrapper/TextWrapper.php [hunk 5]
```diff
@@ -107,6 +110,45 @@ public function getRollback($env = null, $target = null)
...
+        if(is_null($environment)) {
```
Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: I find it strange that you would define the arguments as optional with `= null`, then throw an exception if the **optional** argument is actually null.

File: src/Phinx/Config/Config.php (hunk 1 above)
Comment (alexbowers) [inline src/Phinx/Config/Config.php]: I added it for the PhinxApplication to use, since the PHINX_VERSION wasn't visible. I'll revert these changes as per your other message though.

Comment (shadowhand) [inline src/Phinx/Config/Config.php]: It's injected by `app/phinx.php`.

File: src/Phinx/Wrapper/TextWrapper.php (hunk 3 above)
Comment (alexbowers) [inline src/Phinx/Wrapper/TextWrapper.php]: This was done so that we could predict what is being passed through. I believe that these are the only possible values. Would it be better to accept any, and simply only use these three then?

Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: It's better to accept an array, and then have a `getOption` method, like so:

```php
protected function getOption($name)
{
    if (!isset($this->options[$name])) {
        return null;
    }
    return $this->options[$name];
}
```

That way an option can be added in the future without having to modify code in two places.

File: src/Phinx/Wrapper/TextWrapper.php (hunk 4 above)
Comment (alexbowers) [inline src/Phinx/Wrapper/TextWrapper.php]: This was so that we have an alternative option in setting the environment, based on user preference. They could either write it as this:
```php
<?php
$app = new PhinxApplication();
$wrap = new TextWrapper($app, array(
    ...
));
```

Or it could be done by:

```php
<?php
$app = new PhinxApplication();
$wrap = new TextWrapper($app);
$wrap->setEnvironment('Environment Name');
```

They would both do the same thing, but its just a different way for the user to write it.

File: src/Phinx/Wrapper/TextWrapper.php (hunk 5 above)
Comment (alexbowers) [inline src/Phinx/Wrapper/TextWrapper.php]: Yeah, im not sure why I did that, I'll just make it a required argument.

File: src/Phinx/Wrapper/TextWrapper.php [hunk 6]
```diff
@@ -71,7 +74,7 @@ public function getExitCode()
      */
     public function getStatus($env = null)
     {
-        $command = ['status', '-e' => $env ?: $this->env];
```
Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: Which then makes code like this easier:

```php
$command = ['status', '-e' => $env ?: $this->getOption('environment')];
```

File: src/Phinx/Config/Config.php (hunk 1 above)
Comment (alexbowers) [inline src/Phinx/Config/Config.php]: Would it not make sense to have it at a deeper level than there?

File: src/Phinx/Wrapper/TextWrapper.php (hunk 3 above)
Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: Also, default options can be defined like so:

```php
$options += array(
    'environment' => 'development',
);
```

File: src/Phinx/Config/Config.php (hunk 1 above)
Comment (shadowhand) [inline src/Phinx/Config/Config.php]: No, I don't think so. Dependency Inversion implies that the more dynamic or configurable something is, the **shallower** it should be. See #413 for additional detail on this topic.

Comment (alexbowers) [inline src/Phinx/Config/Config.php]: Okay, I'll remove this. To me it seems messy to have to define PhinxApplication with a version number though, for what seems like purely to be used for the initial output line to report the version. That number can't be trusted much can it?

Comment (shadowhand) [inline src/Phinx/Config/Config.php]: Nothing can be trusted, this is programming. ;) FWIW, that's just how [Symfony Console Application works](http://api.symfony.com/2.4/Symfony/Component/Console/Application.html#method___construct).

**Comment by alexbowers** [comment]
@shadowhand Changes made. Is this better?

File: src/Phinx/Wrapper/TextWrapper.php [hunk 7]
```diff
@@ -107,6 +113,30 @@ public function getRollback($env = null, $target = null)
...
+	 */
+    protected function getOption($key)
+    {
+        if(!isset($this->options[$key])) {
+	        return null;
+        }
+	    return $this->options[$key];
+	}
+
+	/**
```
Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: Indentation is all screwy here.

File: src/Phinx/Wrapper/TextWrapper.php [hunk 8]
```diff
@@ -107,6 +113,30 @@ public function getRollback($env = null, $target = null)
...
+	 * Set option in options array
+	 *
+	 * @param $key
```
Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: Missing type hint and return in this docblock.

File: src/Phinx/Wrapper/TextWrapper.php [hunk 9]
```diff
@@ -107,6 +113,30 @@ public function getRollback($env = null, $target = null)
...
+	 * @param $value
+	 */
+	public function setOption($key, $value) {
+		$this->options[$key] = $value;
+	}
```
Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: Ideally, setters will `return $this`, for fluent usage.

File: src/Phinx/Wrapper/TextWrapper.php [hunk 10]
```diff
@@ -41,18 +41,23 @@
...
+     * @param PhinxApplication $app
+     * @param array $options
      */
-    public function __construct(PhinxApplication $app, $default_env = 'development')
+    public function __construct(PhinxApplication $app, Array $options = array())
     {
+        $options += array(
+            'environment' => 'development',
+            'config_path' => './phinx.yml',
+            'parser' => 'yaml'
```
Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: No need to set parser, it is automatically determined by config file extension.

File: src/Phinx/Wrapper/TextWrapper.php [hunk 11]
```diff
@@ -41,18 +41,23 @@
...
-     * @return void
+     * @param PhinxApplication $app
+     * @param array $options
      */
-    public function __construct(PhinxApplication $app, $default_env = 'development')
+    public function __construct(PhinxApplication $app, Array $options = array())
     {
+        $options += array(
+            'environment' => 'development',
+            'config_path' => './phinx.yml',
```
Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: It would be better if this option name matched the command line name "configuration".

**Comment by alexbowers** [comment]
@shadowhand I think this is all the changes made. Any other thoughts?

File: src/Phinx/Wrapper/TextWrapper.php [hunk 12]
```diff
@@ -40,88 +40,119 @@
...
-        $this->exit_code = $this->app->doRun(new ArrayInput($command), new StreamOutput($stream));
-
-        // Get the output of the command and close the stream, which will
-        // destroy the temporary file.
-        $result = stream_get_contents($stream, -1, 0);
-        fclose($stream);
-
-        return $result;
-    }
+	private $app;
```
Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: You replaced all space indents with tab indents, which violates PSR style. Please fix.

File: src/Phinx/Wrapper/TextWrapper.php [hunk 13]
```diff
@@ -40,88 +40,119 @@
...
+	private $options;
+	private $exit_code;
+
+	/**
+	 * @param PhinxApplication $app
+	 * @param array $options
+	 */
+	public function __construct(PhinxApplication $app, Array $options = array())
```
Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: `Array` type hint should be lowercase.

Comment (alexbowers) [inline src/Phinx/Wrapper/TextWrapper.php]: Elsewhere it is uppercase too 
```php
protected function executeRun(Array $command)
	{
		// Output will be written to a temporary stream, so that it can be
		// collected after running the command.
		$stream = fopen('php://temp', 'w+');

		// Execute the command, capturing the output in the temporary stream
		// and storing the exit code for debugging purposes.
		$this->exit_code = $this->app->doRun(new ArrayInput($command), new StreamOutput($stream));

		// Get the output of the command and close the stream, which will
		// destroy the temporary file.
		$result = stream_get_contents($stream, -1, 0);
		fclose($stream);

		return $result;
	}
```
(TextWrapper.php:141)

Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: That's my mistake from when I wrote the original. Everywhere else in Phinx is lowercase. Doesn't really matter, just good to be consistent.

Comment (alexbowers) [inline src/Phinx/Wrapper/TextWrapper.php]: Want me to change yours too while I'm at it?

Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: Yes please.

File: src/Phinx/Wrapper/TextWrapper.php (hunk 12 above)
Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: Tabs still need to be replaced.

**Comment by alexbowers** [comment]
@shadowhand All of the spaces should be there now. I think it was an issue with me making an edit with Githubs edit feature since it was set to tab indent by default for me. This should be correct now since I have fixed it in phpstorm.

File: src/Phinx/Wrapper/TextWrapper.php [hunk 14]
```diff
@@ -1,127 +1,182 @@
...
-{
-    private $app;
-    private $env;
-    private $exit_code;
-
-    /**
-     * @param  Phinx\Console\PhinxApplication $app
-     * @param  String                         $default_env  environment fallback, defaults to "development"
-     * @return void
+    /* Phinx
```
Comment (shadowhand) [inline src/Phinx/Wrapper/TextWrapper.php]: Now all this is indented incorrectly... please review your own changes before pushing to Github.

**Comment by alexbowers** [comment]
@shadowhand Sorry about the indenting. Fixed it now
